### PR TITLE
Issue #386: Create Docker Compose - front & back end

### PR DESCRIPTION
**Pull request by jahabeebs**
relates to #386 

This PR removes the local file references from the examples Dockerfile (it still runs successfully on its own though) and creates a draft Docker Compose that spins up both the new eidolon-ui2 front end and eidolon-server. I have some questions on how to move forward with it 

## Instructions to run
1) Add .env to root of repo w/OPENAI_API_KEY set
2) run ```docker-compose up``` from the eidolon root directory

## Questions

Both the front end and back end are building and running fine, but when I interact with the running front end it seems to not be communicating with the back end correctly. Some server calls are successful, like this one when interacting with the chatbot:

```bash
eidolon-server_1  | INFO - Request: GET http://eidolon-server:8080/processes?skip=0&limit=100
eidolon-server_1  | INFO - Request: GET http://eidolon-server:8080/openapi.json
eidolon-server_1  | INFO - Response: 200
```

but others are not, like this chatbot one:

```bash
eidolon-server_1  | INFO - Request: POST http://eidolon-server:8080/processes
eidolon-server_1  | INFO - Agent conversational_agent does not exist
eidolon-server_1  | INFO - Response: 404

```

Is my problem that I should be running the usage-server as well as the eidolon-server? Should I not be running eidolon-server from the examples Dockerfile? Do I need to add each example individually to the Dockerfile to get each of them to work? In this case I'll probably just start with one and then start adding end-to-end tests.

I also noticed that the examples Dockerfile has command ```CMD ["resources"]``` but when I added it to the Docker compose like this it always failed to recognize resources and the server crashed immediately, so I removed it. This was the failing part of the docker-compose.yml that I removed:

```yml
  eidolon-server:
    build:
      context: ./examples
      dockerfile: Dockerfile
    environment:
      - PYTHONUNBUFFERED=1
      - OPENAI_API_KEY=${OPENAI_API_KEY}
    expose:
      - "8080"
    ports:
      - "8080:8080"
    volumes:
      - ./examples/resources:/app/resources
    command: ["eidolon-server", "resources"]
```

### CODE REVIEW

**Comment by jahabeebs** [comment]
I can't add you as reviewers so I'll mention you here 😁 @dbrewster @LukeLalor

**Comment by jahabeebs** [comment]
> I can't add you as reviewers so I'll mention you here 😁 @dbrewster @LukeLalor

Update: I talked with Dave about getting the chatbot example running first so I made changes to the docker-compose.yml to use that config for eidolon-server. Now I'm not getting a 404 error when trying to create a new chat but I am getting a 401...specifically this:
```
eidolon-eidolon-server-1  | INFO - Request: POST http://eidolon-server:8080/processes
eidolon-eidolon-server-1  | INFO - Auth Denied: Authorization header missing
eidolon-eidolon-server-1  | INFO - Response: 401
```

When I log in I just use any email it redirects me back to the homepage and shows I'm signed in, and I get a 200 response when we get the session data, so I don't know why it doesn't want to run processes. Should I be running something in the docker-compose.yml that isn't already running? I noticed there are some mentions of "processes" in the sdk Dockerfile and eidolon-server is the entry point, should I be incorporating that Docker file into my docker-compose maybe?

**Comment by LukeLalor** [comment]
The problem is that the dockerfile built enables authorization when building the docker images.

`COPY demo_resources/ resources/`

we should probably build the images without that line for now and think about how to configure this dynamically better

**Comment by jahabeebs** [comment]
> The problem is that the dockerfile built enables authorization when building the docker images.
> 
> `COPY demo_resources/ resources/`
> 
> we should probably build the images without that line for now and think about how to configure this dynamically better

@LukeLalor Thanks for the advice, I excluded the auth yaml from the server Dockerfile and it wasn't giving me the 401s anymore but it was giving me errors because I didn't have a MongoDB configuration. I configured MongoDB in the docker-compose and I was able to get it working. I also added a .env.example so that the docker-compose.yml example will work when running locally

chatbot working

My only worry is that I may have broken the Readme.md instructions for running the conversation chatbot on its own? When I start a virtual env and follow the instructions it doesn't like me input and gives me a 422. Maybe it's unrelated though. When I change the curl to be in JSON like this I don't get a 422 anymore but the input is blank:

JSON request (not 422 but blank response)

```curl
curl -X 'POST' "http://localhost:8080/processes/$PROCESS_ID/agent/conversational_agent/actions/converse" -H "Content-Type: application/json" -d '{"input":"What kind of tools can I build with LLM agents?"}'
```

Unsuccessful example request in chatbot Readme

breaking dockerfile?

**Comment by jahabeebs** [comment]
@LukeLalor @dbrewster 

I made the final edits to our chatbot e2e test and the GitHub Action is running successfully. Here's the latest test run on my eidolon fork:

https://github.com/jahabeebs/eidolon/actions/runs/9280656466/job/25535456261

This is a video of what it looks like locally when you run pnpm run playwright:debug

https://github.com/eidolon-ai/eidolon/assets/47253537/951c224a-4160-4541-ae0c-b05914b3af3a

Some configuration things you may or may not want me to change:
1) I have the e2e.yml running on pushes to main and PRs, not sure if this is too much. The e2e test is very heavy and takes around 5-6 minutes to run. This configuration is the same as the test.yml configuration
```yml
on:
  push:
    branches: [main]
  pull_request:
  ```
2) I've configured playwright to upload a screenshot if a failure occurs to webui/apps/eidolon-ui2/tests/screenshots. This way you can see if it's a real failure or if the chatbot is just hallucinating. I can remove this logic if you'd like

In order for e2e.yml to run successfully in the eidolon repo the following secrets need to be set in GitHub:
```yml
      AUTH_TRUST_HOST: ${{ secrets.AUTH_TRUST_HOST }}
      EIDOLON_SERVER: ${{ secrets.EIDOLON_SERVER }}
      NEXTAUTH_SECRET: ${{ secrets.NEXTAUTH_SECRET }}
      NEXT_PUBLIC_API_URL: ${{ secrets.NEXT_PUBLIC_API_URL }}
      NEXT_PUBLIC_DEBUG: ${{ secrets.NEXT_PUBLIC_DEBUG }}
      NEXT_PUBLIC_LOG_LEVEL: ${{ secrets.NEXT_PUBLIC_LOG_LEVEL }}
      OPENAI_API_KEY: ${{ secrets.OPENAI_API_KEY }}
```

File: docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+      - "27017:27017"
+    volumes:
+      - mongo-data:/data/db
+
+  eidolon-server:
+    build:
+      context: ./examples
+      dockerfile: Dockerfile
+      args:
+        - EIDOLON_VERSION=latest
```
Comment (LukeLalor) [inline docker-compose.yml:18]: This will not build the right version unless the latest version has been built locally before hand. Since we are just using the chatbot (which does not have any additional dependencies) It might be worth depending on the dockerfile in sdk. Alternatively, this is probably fine for now if we just add the sdk build to the script

File: webui/apps/eidolon-ui2/package.json [hunk 2]
```diff
@@ -24,6 +24,12 @@
     "dev": "(patch -N ../../node_modules/.pnpm/@jsdevtools+ono@7.1.3/node_modules/@jsdevtools/ono/esm/index.js ono_turbo.patch || true) && npm run build-eidolon && next dev --turbo",
     "build": "npm run build-eidolon && next build",
     "build:static": "npm run build-eidolon && next build && next export",
+    "docker:up": "docker-compose up -d",
+    "docker:down": "docker-compose down",
+    "docker:build": "docker-compose build",
```
Comment (LukeLalor) [inline webui/apps/eidolon-ui2/package.json:29]: this should probably build the dockerfile in sdk as well.

File: .github/workflows/e2e.yml [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+      - name: Build Docker images
+        run: pnpm run docker:build
+        working-directory: ./webui/apps/eidolon-ui2
+
+      - name: Start Docker containers
+        run: pnpm run docker:up
+        working-directory: ./webui/apps/eidolon-ui2
+
+      - name: Print MongoDB logs
+        run: docker logs $(docker ps -q --filter name=mongo)
```
Comment (LukeLalor) [inline .github/workflows/e2e.yml:77]: we probably don't want to print the mongodb logs, but they would be helpful as an artifact

File: .github/workflows/e2e.yml [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+      - name: Run Playwright tests
+        run: |
+          pnpm exec playwright test --config=playwright.config.js
+        working-directory: ./webui/apps/eidolon-ui2
+
+      - name: Stop Docker containers
+        run: pnpm run docker:down
+        working-directory: ./webui/apps/eidolon-ui2
+
+      - name: Upload Playwright test results
```
Comment (LukeLalor) [inline .github/workflows/e2e.yml:91]: I think the upload artifact steps need a if: always() on then to run in case the playwright tests fail.

File: examples/Dockerfile [hunk 5]
```diff
@@ -14,7 +15,8 @@ ARG LOC=eidolon_examples
 FROM examples
 WORKDIR /app
 COPY $LOC .
-COPY demo_resources/ resources/
+COPY demo_resources/usage_resources.yaml resources/
+COPY eidolon_examples/conversational_chatbot/resources/conversational_agent.yaml /app/resources/conversational_agent.yaml
```
Comment (LukeLalor) [inline examples/Dockerfile:19]: I saw that you are mounting this file in the docker compose, is this addition still needed?

File: examples/Dockerfile [hunk 6]
```diff
@@ -14,7 +15,8 @@ ARG LOC=eidolon_examples
 FROM examples
 WORKDIR /app
 COPY $LOC .
-COPY demo_resources/ resources/
+COPY demo_resources/usage_resources.yaml resources/
```
Comment (LukeLalor) [inline examples/Dockerfile:18]: is this needed for the ui? We should make a todo to remove this requirement so the ui gracefully handles this being missing.

File: examples/Dockerfile [hunk 7]
```diff
@@ -3,8 +3,9 @@ FROM python:3.11-slim as builder
 RUN pip install poetry
 COPY . .
 RUN mkdir dist
-RUN touch dist/requirements.txt
-RUN poetry export --without dev --without-hashes --format=requirements.txt | sed '/@ file/d' > dist/requirements.txt
+
+# Generate requirements.txt without local file references
+RUN poetry export --without dev --without-hashes --format=requirements.txt | sed '/^-e /d' > dist/requirements.txt
```
[the remark this answers is not in the document]
Comment (jahabeebs) [reply examples/Dockerfile]: Ehh it didn't make a difference but I added it back to the sdk Dockerfile. I reverted all my changes from this examples Dockerfile since we won't be using it anyway

File: docker-compose.yml [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+# Run `docker-compose up` in the same directory as this file to start the services.
+version: '3.8'
+
+services:
+  mongo:
+    image: mongo:latest
+    ports:
+      - "27017:27017"
+    volumes:
+      - mongo-data:/data/db
```
[the remark this answers is not in the document]
Comment (jahabeebs) [reply docker-compose.yml:11]: Added mongo-data/ to the gitignore

File: docker-compose.yml (hunk 1 above)
Comment (jahabeebs) [reply docker-compose.yml:18]: I managed to switch to the SDK Dockerfile without a problem. I added the same Dockerfile change as the other:
```
RUN poetry export --without dev --without-hashes --format=requirements.txt | sed '/^-e /d' > dist/requirements.txt
```

File: webui/apps/eidolon-ui2/package.json (hunk 2 above)
Comment (jahabeebs) [reply webui/apps/eidolon-ui2/package.json]: Added/modified the following scripts:
```
    "docker:build-sdk": "docker build -t eidolonai/sdk_base:latest -f ../../../sdk/Dockerfile ../../../sdk/",
    "docker:build": "npm run docker:build-sdk && docker-compose build",
```

File: .github/workflows/e2e.yml (hunk 3 above)
Comment (jahabeebs) [reply .github/workflows/e2e.yml]: Added these instead:
```yml
      - name: Save MongoDB logs
        run: docker logs $(docker ps -q --filter name=mongo) > mongodb-logs.txt
        
      - name: Upload MongoDB logs
        uses: actions/upload-artifact@v3
        if: always()
        with:
          name: mongodb-logs
          path: mongodb-logs.txt
```

File: .github/workflows/e2e.yml (hunk 4 above)
Comment (jahabeebs) [reply .github/workflows/e2e.yml:99]: Good point, I added the always() to the steps after the playwright test

File: examples/Dockerfile (hunk 5 above)
Comment (jahabeebs) [reply examples/Dockerfile]: Since we're using the SDK Dockerfile I just reverted all the changes I made to the examples Dockerfile

File: examples/Dockerfile (hunk 6 above)
Comment (jahabeebs) [reply examples/Dockerfile]: Sort of, the problem was that the auth yaml was not working properly with the front end (that was the cause of the 403 forbidden errors I mentioned earlier). I made that change to only copy over usage_resources.yml so that the front end and back end could communicate properly

**Comment by jahabeebs** [comment]
@LukeLalor I believe I've addressed all the changes, here's the successful actions run with the changes I made based on your comments. Thanks!

https://github.com/jahabeebs/eidolon/actions/runs/9312680168/job/25633823756?pr=1

**Comment by LukeLalor** [comment]
👀

**Review by LukeLalor — APPROVED** [review]
I say ship it. I created a new branch `integration_branch` so we can make sure the tests pass over there.

@jahabeebs can you retarget? Then I will force merge

**Comment by jahabeebs** [comment]
> I say ship it. I created a new branch `integration_branch` so we can make sure the tests pass over there.
> 
> @jahabeebs, let me know if you are unable to and I can force merge it.

@LukeLalor I tried but don't have permissions, could you force merge?

**Comment by LukeLalor** [comment]
@jahabeebs done, but the tests are failing over on the integration branch
